### ISSUE TITLE
map::Just cannot be changed to Nothing

### ISSUE THREAD

**Issue by GheorgheP** [jlmorgan/node-lodash-fantasy#28]
If the `map` callback return and undefined, map return an `Nothing` instead of `Just(undefined)`

E.g.:

```js
Maybe.from(2).map(_ => undefined) // => Nothing()
```
But it should be:

```js
Maybe.from(2).map(_ => undefined) // => Just(undefined)
```

`map` cannot modify the monad, it modifies the inside value only.

Currently `map` behaviors like `chain`, but it should not.

**Comment by jlmorgan** [comment]
This is true in the context of a language where `Nothing` is synonymous with `undefined` and `null`, but because JavaScript doesn't support this concept, you have to do a check on whether or not the returned value _is_ of value. Where as in other languages, such a Haskel, you _have_ to return _something_; otherwise, as you stated, it's more of a `Chain#chain` or `Monad#bind` behavior or similar to `Promise#then`. If you look at other languages where `null` is a concept, they implement the same behavior as a trap and to prevent breaking/throwing exceptions for `NullPointerException`.

The whole point of the `Maybe` is to provide that guard around `null` and `undefined` because they are the absence of value. If you allow `Functor#map` in the context of a `Maybe` to return `null` or `undefined`, then you next mapping function has to be `null`/`undefined` sensitive which defeats the purpose of the `Maybe` from the start.

Now, in the context of `Either`, `Validation`, or the base `Identity`, they are relying on the functions to do all of the work. That's why you will often see something like `Either<Error, Maybe<T>>` because they want that inner value to be safely guarded through the construct.

Thoughts?

**Comment by GheorgheP** [comment]
Yes, I had this thought too, before having situations with `null`. You see the case with `undefined` is clear, `undefined` may be called `bottom` in Haskell (it's just an thought), it really means nothing, but the problem with `null` is that sometimes it is tread as value, I mean cases where `null` and `undefined` aren't the same, `null` being a value and `undefined` a non-value. I reported this because I thought is't better to provide explicitly the `map` functionality and to give the developer ability to use `chain` instead of map knowing that his function may or not return `undefined`, against to the fact that this work will be done implicitly, and this being way harder to handle.

But, as you mentioned in another [issue](https://github.com/jlmorgan/node-lodash-fantasy/issues/26#issuecomment-317029590) that his project is designed for beginners that are really starting on functional programming, this approach may be the right one, as I had this issue few months ago with `ramda-fantasy` and initially I cold not understand why `map'  doesn't handle **nullable** values, thanks to [Slack functional programming](https://functionalprogramming.slack.com/) community, helping me with that.

**Conclusion**: I think you got the point and it may make life easier for many developers, especially beginners in fp.

**Comment by jlmorgan** [comment]
I had original thought about the approach of doing `undefined` as synonymous with `Nothing`, but then it brought in the question of how to deal with `null`. Since languages like Java and C# don't have the concept of `undefined`, but only `null`, I decided to treat the two as the same.

**Comment by GheorgheP** [comment]
Hmmmm, I think you have to choose between two "evils" then.

Currently:

```js
Maybe.Just.from(1).map(_ => undefined) // >_  Nothing()
```
But

```js
Maybe.Just.from() // >_ Just { value: undefined }
```
So you nee to decide the workflow, to parse **nullables** implicitly or not, as at the moment it violates the _associativity_ principle  . Choosing different approaches, you get different values, that not being right for sure.

**Comment by jlmorgan** [comment]
You are correct here. I'm in the processing of addressing this in my refactor. Part it came from not really knowing what I wanted to do about certain situations.
